controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
 
-	if d, found := a.instance.Status.Decisions[subscription.Name]; found {
-		mw := &ocmworkv1.ManifestWork{}
+	// TODO: implement relocation
+	return true, nil
+}
 
-		err := a.reconciler.Get(a.ctx, types.NamespacedName{Name: mwName, Namespace: d.HomeCluster}, mw)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				return nil
-			}
+func (a *AVRInstance) isAlreadyFailedBack(targetCluster string) bool {
+	if len(a.userPlacementRule.Status.Decisions) > 0 &&
+		targetCluster == a.userPlacementRule.Status.Decisions[0].ClusterName {
+		a.log.Info("Already failed back", "state", a.instance.Status.LastKnownDRState)
 
-			return fmt.Errorf("failed to retrieve manifestwork for type: %s. Error: %w", mwName, err)
-		}
+		return true
+	}
+
+	return false
+}
 
-		a.log.Info("deleting ManifestWork", "name", mw.Name)
+func (a *AVRInstance) getPreferredClusterNamespaced() (string, string) {
+	preferredCluster := a.instance.Spec.PreferredCluster
+	preferredClusterNamespace := preferredCluster
 
-		return a.reconciler.Delete(a.ctx, mw)
+	if preferredCluster == "" {
+		if a.instance.Status.PreferredDecision != (plrv1.PlacementDecision{}) {
+			preferredCluster = a.instance.Status.PreferredDecision.ClusterName
+			preferredClusterNamespace = a.instance.Status.PreferredDecision.ClusterNamespace
+		}
 	}
 
-	return nil
+	return preferredCluster, preferredClusterNamespace
 }
 
-func (a *AVRInstance) restorePVFromBackup(
-	subscription *subv1.Subscription, homeCluster string) error {
-	a.log.Info("Restoring PVs to new managed cluster", "name", homeCluster)
+func (a *AVRInstance) processVRGForSecondaries(preferredCluster, curHomeCluster string) (bool, error) {
+	const ensured = true
 
-	pvList, err := a.listPVsFromS3Store(subscription)
+	err := a.updateVRGStateToSecondary(curHomeCluster)
 	if err != nil {
-		return errorswrapper.Wrap(err, "failed to retrieve PVs from S3 store")
+		a.log.Error(err, "Failed to update VRG to secondary")
+
+		return !ensured, err
 	}
 
-	a.log.Info(fmt.Sprintf("Found %d PVs for subscription %s", len(pvList), subscription.Name))
+	// During failover, we have already updated VRG replication state for the
+	// preferred cluster to secondary. All we have to do at this point is to
+	// check whether the VRG has already transitioned to secondary. This call
+	// uses ManagedClusterView to check the VRG on the ManagedCluster.
+	secondary := a.ensureVRGIsSecondary(preferredCluster)
+	if !secondary {
+		a.log.Info("Still waiting for VRG to transition to secondary", "clusterName", preferredCluster)
 
-	if len(pvList) == 0 {
-		return nil
+		return !ensured, nil
+	}
+
+	// Should we make sure that the current home cluster has already transitioned to secondary
+	// before we proceed with the failback? Is setting the ManifestWork VRG to secondary is enough?
+	// Nevertheless, we are checking here.  If this is an overkill, we can remove it.
+	secondary = a.ensureVRGIsSecondary(curHomeCluster)
```

Comment:
As Subscriptions are still placed on curHomeCluster, VRG will not report Secondary. For VRG to report Secondary on curCluster the Subscriptions (and hence the PVCs) would need to be removed. This would need setting the userPlacement status to no clusters and then ensuring VRG is secondary.

Reply:
So what you are saying is that the VRG state will not transition to secondary unless all the PVCs have been deleted.  But that is not right. Why are we setting the VRG state to secondary then? Why don't we just delete the VRG after we clear the placement rule status? And internally, the VRG transitions to secondary then it removes its finalizer?

Reply:
Done